Date: 2005-08-07
Initial Package Version: 2.9.6
Upstream Status: Not yet submitted
Origin: Miklos Vajna <vmiklos@frugalware.org>
Description: Allow users to view the changelog of a package.
 doc/pacman.8.in |    3 +++
 scripts/makepkg |   18 ++++++++++++++----
 src/pacman.c    |   37 +++++++++++++++++++++++++++++++++++--
 3 files changed, 52 insertions(+), 6 deletions(-)
diff -Naur pacman-2.9.6.orig/doc/pacman.8.in pacman-2.9.6/doc/pacman.8.in
--- pacman-2.9.6.orig/doc/pacman.8.in	2005-08-09 14:04:21.000000000 +0200
+++ pacman-2.9.6/doc/pacman.8.in	2005-08-09 14:04:41.000000000 +0200
@@ -157,6 +157,9 @@
 This option is analagous to a backwards --sync operation.
 .SH QUERY OPTIONS
 .TP
+.B "\-c, \-\-changelog"
+View the changelog of a package.
+.TP
 .B "\-e, \-\-orphans"
 List all packages that were explicitly installed (ie, not pulled in
 as a dependency by other packages) and are not required by any other
diff -Naur pacman-2.9.6.orig/scripts/makepkg pacman-2.9.6/scripts/makepkg
--- pacman-2.9.6.orig/scripts/makepkg	2005-08-09 14:04:21.000000000 +0200
+++ pacman-2.9.6/scripts/makepkg	2005-08-09 14:04:41.000000000 +0200
@@ -923,15 +923,25 @@
 cd $startdir/pkg
 tar cvf /dev/null * | sort >.FILELIST
 
+# changelog
+if darcs --commands 2>&1|grep -q add; then
+	msg "Generating .CHANGELOG file..."
+	cd $startdir
+	darcs changes $BUILDSCRIPT |sed 's|source/||;s|/FrugalBuild||' \
+		>pkg/.CHANGELOG
+fi
+
 # tar it up
 msg "Compressing package..."
 cd $startdir/pkg
 if [ -f $startdir/pkg/.INSTALL ]; then
-	cmd="tar czvf $PKGDEST/$pkgname-$pkgver-$pkgrel-$CARCH.fpm .PKGINFO .FILELIST .INSTALL *"
-else
-	cmd="tar czvf $PKGDEST/$pkgname-$pkgver-$pkgrel-$CARCH.fpm .PKGINFO .FILELIST *"
+	extra=".INSTALL"
+fi
+if [ -f $startdir/pkg/.CHANGELOG ]; then
+	extra="$extra .CHANGELOG"
 fi
-$cmd | sort >../filelist
+tar czvf $PKGDEST/$pkgname-$pkgver-$pkgrel-$CARCH.fpm .PKGINFO .FILELIST \
+	$extra * | sort >../filelist
 
 cd $startdir
 if [ "$SEARCHDEPS" == "1" ]; then
diff -Naur pacman-2.9.6.orig/src/db.c pacman-2.9.6/src/db.c
--- pacman-2.9.6.orig/src/db.c	2005-08-09 14:04:21.000000000 +0200
+++ pacman-2.9.6/src/db.c	2005-08-09 14:04:41.000000000 +0200
@@ -629,6 +629,9 @@
 	/* INSTALL */
 	snprintf(path, PATH_MAX, "%s/install", topdir);
 	unlink(path);
+	/* CHANGELOG */
+	snprintf(path, PATH_MAX, "%s/changelog", topdir);
+	unlink(path);
 	/* directory */
 	rmdir(topdir);
 
diff -Naur pacman-2.9.6.orig/src/pacman.c pacman-2.9.6/src/pacman.c
--- pacman-2.9.6.orig/src/pacman.c	2005-08-09 14:04:21.000000000 +0200
+++ pacman-2.9.6/src/pacman.c	2005-08-09 14:08:14.000000000 +0200
@@ -72,6 +72,7 @@
 unsigned short pmo_q_orphans    = 0;
 unsigned short pmo_q_owns       = 0;
 unsigned short pmo_q_search     = 0;
+unsigned short pmo_q_changelog  = 0;
 unsigned short pmo_r_cascade    = 0;
 unsigned short pmo_r_dbonly     = 0;
 unsigned short pmo_r_recurse    = 0;
@@ -1808,10 +1809,17 @@
 				continue;
 			}
 
-			if(!strcmp(pathname, "._install") || !strcmp(pathname, ".INSTALL")) {
-				/* the install script goes inside the db */
-				snprintf(expath, PATH_MAX, "%s%s/%s/%s-%s/install", pmo_root,
-									pmo_dbpath, db->treename, info->name, info->version);
+			if(!strcmp(pathname, "._install") || !strcmp(pathname, ".INSTALL") || !strcmp(pathname, ".CHANGELOG")) {
+				if(!strcmp(pathname, ".CHANGELOG")) {
+					/* the changelog goes inside the db */
+					snprintf(expath, PATH_MAX, "%s%s/%s/%s-%s/changelog", pmo_root,
+										pmo_dbpath, db->treename, info->name, info->version);
+				} else {
+					/* the install script goes inside the db */
+					snprintf(expath, PATH_MAX, "%s%s/%s/%s-%s/install", pmo_root,
+										pmo_dbpath, db->treename, info->name, info->version);
+				}
+			
 			} else {
 				/* build the new pathname relative to pmo_root */
 				snprintf(expath, PATH_MAX, "%s%s", pmo_root, pathname);
@@ -2514,12 +2522,36 @@
 			}
 		} else {
 			/* find a target */
-			if(pmo_q_info || pmo_q_list) {
+			if(pmo_q_changelog || pmo_q_info || pmo_q_list) {
 				info = db_scan(db, package, INFRQ_ALL);
 				if(info == NULL) {
 					fprintf(stderr, "Package \"%s\" was not found.\n", package);
 					return(2);
 				}
+				if(pmo_s_clean)
+				{
+					char clfile[PATH_MAX];
+					FILE* fp = NULL;
+					char line[PATH_MAX+1];
+					snprintf(clfile, PATH_MAX, "%s%s/%s/%s-%s/changelog", pmo_root, pmo_dbpath, db->treename, info->name, info->version);
+
+					if((fp = fopen(clfile, "r")) == NULL)
+					{
+						perror(clfile);
+						return(1);
+					}
+					else
+					{
+						while(!feof(fp))
+						{
+							fgets(line, PATH_MAX, fp);
+							printf("%s", line);
+							line[0] = '\0';
+						}
+					}
+					fclose(fp);
+					return(0);
+				}
 				if(pmo_q_info) {
 					dump_pkg_full(info);
 					if(pmo_q_info > 1 && info->backup) {
@@ -3432,6 +3464,7 @@
 		{"dbpath",     required_argument, 0, 'b'},
 		{"clean",      no_argument,       0, 'c'},
 		{"cascade",    no_argument,       0, 'c'},
+		{"changelog",  no_argument,       0, 'c'},
 		{"nodeps",     no_argument,       0, 'd'},
 		{"orphans",    no_argument,       0, 'e'},
 		{"force",      no_argument,       0, 'f'},
@@ -3478,7 +3511,7 @@
 			case 'h': pmo_help = 1; break;
 			case 'V': pmo_version = 1; break;
 			case 'b': strcpy(pmo_dbpath, optarg); break;
-			case 'c': pmo_s_clean++; pmo_r_cascade = 1; break;
+			case 'c': pmo_s_clean++; pmo_r_cascade = 1; pmo_q_changelog = 1; break;
 			case 'd': pmo_nodeps = 1; break;
 			case 'e': pmo_q_orphans = 1; break;
 			case 'f': pmo_force = 1; break;
@@ -3803,6 +3836,7 @@
 		} else if(op == PM_QUERY) {
 			printf("usage:  %s {-Q --query} [options] [package]\n", myname);
 			printf("options:\n");
+			printf("  -c, --changelog     view the changelog of a package\n");
 			printf("  -e, --orphans       list all packages that were explicitly installed\n");
 			printf("                      and are not required by any other packages\n");
 			printf("  -g, --groups        view all members of a package group\n");
